README.md
# Testing utilities for RethinkDB and AVA

[![ava-rethinkdb on NPM](https://img.shields.io/npm/v/ava-rethinkdb.svg)](https://npm.im/ava-rethinkdb)
[![Build Status](https://travis-ci.org/rrdelaney/ava-rethinkdb.svg?branch=master)](https://travis-ci.org/rrdelaney/ava-rethinkdb)
![See package.json for Node.js Support](https://img.shields.io/node/v/ava-rethinkdb.svg?label=Runs%20on%20Node.js)

```
npm install --save-dev ava-rethinkdb
```

This is a hacky way of using the NodeJS RethinkDB and AVA together. It uses
undocumented features of the RethinkDB driver, and should be considered
experimental.

# Basic Testing

By running `init` and `cleanup` you get a fully managed database instance for
your tests! Everything is cleaned up at the end, so there's no leftover fixtures

```js
import test from 'ava'
import { init, cleanup } from 'ava-rethinkdb'

test.before(init())
test.after.always(cleanup)

test('I should have a RethinkDB instance', async t => {
  let connection = await r.connect({})

  await r.dbCreate('MyDatabase')
})
```

# Seeding

The problem is that if you want to do multiple tests they all happen at the same
time due to the magic of AVA. Luckily you can seed the database with a simple
JSON structure

```js
import test from 'ava'
import { init, cleanup } from 'ava-rethinkdb'

const TEST_DATA = {
  my_database: { // The top level is the database to create
    my_table: [ // Next is a table in the database. This holds an array of documents to insert
      { name: 'A', value: 1},
      { name: 'B', value: 2}
    ],
    users: [
      { username: 'daniel', email: 'wry@gmail.com' },
      { username: 'heya', email: 'ayeh@outlook.com' }
    ]
  }
}

test.before(init(TEST_DATA))
test.after.always(cleanup)

test('These documents should exist', async t => {
  let conn = await r.connect({ db: 'my_database' })
  let results = await r.table('my_table').run(conn)
  let data = await results.toArray()

  console.log(data)
  t.truthy(data)
})
```

# Different Database Instances

This is where the magic really is. Every single test file is given its own
RethinkDB instance. This makes it perfect for integration tests against
endpoints, because now they can all be used in parallel! The magic comes
from modifying the default port the driver looks at, making it different
in each process, then spinning up a RethinkDB instance at that port.
Check out the `test` directory for a good example.

```js
// app.js

const express = require('express')
const r = require('rethinkdb')

let app = express()

app.get('/users', (req, res) => {
  r.connect({ db: 'app' })
    .then(conn => r.table('users').run(conn))
    .then(results => results.toArray())
    .then(users => res.status(200).send({ users }))
    .catch(e => res.status(500).send(e))
})

module.exports = { app }
```

```js
// test/integration/users-test-1.js
import test from 'ava'
import request from 'supertest-as-promised'
import { init, cleanup } from 'ava-rethinkdb'

import { app } from '../../app.js'

const TEST_DATA = {
  app: {
    users: [
      { name: 'UserA' },
      { name: 'UserB' }
    ]
  }
}

test('Users should be returned from /users', t => {
  return request(app)
    .get('/users')
    .expect(200)
})
```

```js
// test/integration/users-test-2.js
import test from 'ava'
import request from 'supertest-as-promised'
import { init, cleanup } from 'ava-rethinkdb'

import { app } from '../../app.js'

const TEST_DATA = {
  app: {
    users: [
      { name: 'UserC' },
      { name: 'UserD' }
    ]
  }
}

test('Different users should be returned from /users', t => {
  return request(app)
    .get('/users')
    .expect(200)
})
```

The `TEST_DATA` contained in each file creates a new database to be used for
each file!

# Debugging

To view the output from all the server logs, set the environment variable
`AVA_RETHINKDB_DEBUG=on`


CHANGELOG.md
# Changelog
[ava-rethinkdb]: https://github.com/rrdelaney/ava-rethinkdb
[Travis CI]: https://travis-ci.org/

## [Unreleased]
[Unreleased]: https://github.com/rrdelaney/ava-rethinkdb/compare/v0.1.4...HEAD

### Added

- `engines.node = ">=6.0.0"` to formally note which versions of Node are supported.

## [0.1.4] - 2017-08-01

[Code][0.1.4] ([Diff][0.1.4-diff]) | [Changelog][0.1.4-log]

Updates and Bugfixes

### Added

- [Travis CI][] configuration (#5)
- Tests running [ava-rethinkdb][] from multiple files and directories (#7)

### Modified

- Updated AVA to v0.21.0 (#5)
- Ensured data directories are unique to prevent collisions (#6)

[0.1.4]: https://github.com/rrdelaney/ava-rethinkdb/tree/v0.1.4
[0.1.4-diff]: https://github.com/rrdelaney/ava-rethinkdb/compare/141c3414bfd8f3b5f73b54e2630d7b25465c2fd8...v0.1.4
[0.1.4-log]:  https://github.com/rrdelaney/ava-rethinkdb/blob/master/CHANGELOG.md#014---2017-08-01


test/initialize.js
import test from 'ava'
import r from 'rethinkdb'
import { init, cleanup } from '../'

const TEST_DATA = {
  fp: {
    users: [
      { name: 'UserA' },
      { name: 'UserB' }
    ]
  },
  meta: {
    data: [
      { something: true }
    ]
  },
  test: {
    data: [
      { something: true }
    ]
  }
}

test.before('Initialize DB', init(TEST_DATA))
test.after.always('Teardown DB', cleanup)

async function shouldDatabaseExist (t, input, expected) {
  let conn = await r.connect({})
  let databases = await r.dbList().run(conn)

  if (expected) {
    t.true(databases.indexOf(input) !== -1)
  } else {
    t.true(databases.indexOf(input) === -1)
  }
}

shouldDatabaseExist.title = (_, input, expected) => `Database ${input} should${expected ? '' : "n't"} exist`

test(shouldDatabaseExist, 'fp', true)
test(shouldDatabaseExist, 'meta', true)
test(shouldDatabaseExist, 'test', true)
test(shouldDatabaseExist, 'sugar', false)

async function shouldUserExist (t, name, expected) {
  let conn = await r.connect({ db: 'fp' })
  let [ user ] = await r.table('users').filter({ name }).run(conn).then(_ => _.toArray())

  if (expected) {
    t.truthy(user, `${name} should exist`)
  } else {
    t.falsy(user, `${name} should not exist`)
  }
}

shouldUserExist.title = (_, input, expected) => `${input} should${expected ? '' : "n't"} exist`

test(shouldUserExist, 'UserA', true)
test(shouldUserExist, 'UserB', true)
test(shouldUserExist, 'UserC', false)
test(shouldUserExist, 'UserD', false)


test/port-offset.js
import test from 'ava'
import { getPortOffset } from '../'

test('port offsets shouldn\'t exceed the maximum offset', async t => {
  const maxOffset = 65535 - 28015

  t.true(getPortOffset(0) <= maxOffset)
  t.true(getPortOffset(10) <= maxOffset)
  t.true(getPortOffset(maxOffset / 2) <= maxOffset)
  t.true(getPortOffset(maxOffset - 1) <= maxOffset)
  t.true(getPortOffset(maxOffset + 1) <= maxOffset)
  t.true(getPortOffset(maxOffset) <= maxOffset)
  t.true(getPortOffset(maxOffset * 2) <= maxOffset)
  t.true(getPortOffset(maxOffset * 2.5) <= maxOffset)
})


test/not-duplicated.js
import test from 'ava'
import r from 'rethinkdb'
import { init, cleanup } from '../'

const TEST_DATA = {
  fp: {
    users: [
      { name: 'UserC' },
      { name: 'UserD' }
    ]
  }
}

test.before('Initialize DB', init(TEST_DATA))
test.after.always('Teardown DB', cleanup)

async function shouldDatabaseExist (t, input, expected) {
  let conn = await r.connect({})
  let databases = await r.dbList().run(conn)

  if (expected) {
    t.true(databases.indexOf(input) !== -1)
  } else {
    t.true(databases.indexOf(input) === -1)
  }
}

shouldDatabaseExist.title = (_, input, expected) => `Database ${input} should${expected ? '' : "n't"} exist`

test(shouldDatabaseExist, 'fp', true)
test(shouldDatabaseExist, 'meta', false)
test(shouldDatabaseExist, 'sugar', false)

async function shouldUserExist (t, name, expected) {
  let conn = await r.connect({ db: 'fp' })
  let [ user ] = await r.table('users').filter({ name }).run(conn).then(_ => _.toArray())

  if (expected) {
    t.truthy(user, `${name} should exist`)
  } else {
    t.falsy(user, `${name} should not exist`)
  }
}

shouldUserExist.title = (_, input, expected) => `${input} should${expected ? '' : "n't"} exist`

test(shouldUserExist, 'UserA', false)
test(shouldUserExist, 'UserB', false)
test(shouldUserExist, 'UserC', true)
test(shouldUserExist, 'UserD', true)


MockedRethinkDB.js
const filter = require("lodash.filter");
const map = require("lodash.map");
const flatMap = require("lodash.flatmap");
const rimraf = require("rmfr");
const tmp = require("tmp");

/**
 * @typedef {Object} MockedRethinkDBOptions
 * @property {Number} [port] the port to use for RethinkDB.  If port already taken, will try to use other ports.
 * @property {Object} [initialData] initial data to populate RethinkDB with.
 * @property {String} [internalContext] the `t.context` variable to use for the {@link MockedRethinkDB} instance.
 * @todo Document `initialData` (write more `typedef`s)
*/

/**
 * @todo Listen to `opts.internalContext` for static methods.
*/
class MockedRethinkDB {

  /**
   * Fully mock RethinkDB.
   * @param {MockedRethinkDBOptions} [opts] options to control the spawned instance.
   * @return {Function} to be passed to `test.beforeEach()`.
  */
  static init(opts) {
    return initRethinkDB = t => {
      const mocked = new this(opts);
      t.context._mockedRethinkDB = mocked;
      return mocked
        .spawn()
        .then(() => mocked.hackModule())
        .then(() => mocked.importData());
    };
  }

  /**
   * Unmock RethinkDB.  Pairs with {@link MockedRethinkDB.init}.
   * @param {MockedRethinkDBOptions} [opts] options used to spawn the instance.
   * @return {Function} to be passed to `test.afterEach()`.
  */
  static cleanup(opts) {
    return cleanupRethinkDB = t => {
      if(!t.context._mockedRethinkDB || !t.context._mockedRethinkDB instanceof this) {
        console.warn(`cleanup() can't find a mocked RethinkDB instance.  Found ${t.context._mockedRethinkDB}.`);
        return;
      }
      return t.context._mockedRethinkDB.cleanup();
    }
  }

  /**
   * Sets up a series of `beforeEach()` and `afterEach()` statements to mock RethinkDB and cleanup after tests are
   * completed.
   * @param {MockedRethinkDBOptions} [opts] options used to spawn the instance.
   * @return {void}
   * @todo Improve names for each `beforeEach()`/`afterEach()` stage.  Potentially key off of a generated "short code"
   *   or use the test title if possible.
  */
  static mockRethink(opts) {
    const test = require("ava");
    opts.internalContext = "_mockedRethinkDB";

    test.beforeEach("create MockedRethinkDB instance", t => {
      t.context[opts.internalContext] = new this(opts);
    });

    test.beforeEach("generate temporary RethinkDB storage", t => {
      return t.context[opts.internalContext].storage();
    });

    test.beforeEach("spawn local RethinkDB server", t => {
      return t.context[opts.internalContext].spawn();
    });

    test.beforeEach("hack the 'rethinkdb' package to use local server", t => {
      return t.context[opts.internalContext].hackModule();
    });

    if(opts.initialData) {
      test.beforeEach("populate the RethinkDB server with data", t => {
        return t.context[opts.internalContext].importData();
      });
    }

    test.afterEach.always("shutdown 'rethinkdb' server", t => {
      return t.context[opts.internalContext].shutdown();
    });

    test.afterEach.always("cleanup temporary RethinkDB storage", t => {
      return t.context[opts.internalContext].removeStorage();
    });
  }

  /**
   * @param {Object} opts common options to mocking methods.
  */
  constructor(opts) {
    this.opts = opts;
  }

  /**
   * The port to assign to the spawned RethinkDB process.
   * @return {Object} `{ offset, port }` the offset and port to use.
   * @depreciated should not be used in anything except spawning RethinkDB, so spawn can try different ports if one is
   *   in use.
   * @todo If passed a port (e.g. port in use and we need another), select the next port to try.
  */
  port() {
    if(this.opts.port && this.opts.offset) {
      return {
        offset: this.opts.offset,
        port: this.opts.port,
      };
    }
    const maxOffset = 65535 - 28015;
    const pid = process.pid;
    const offset = pid - (Math.floor(pid / maxOffset) * maxOffset);
    return { offset, port: 28015 + offset };
  }

  /**
   * Creates a temporary storage directory to contain the data for the mock database.
   * @return {Promise<String>} resolves with the path for the storage directory once created.
  */
  storage() {
    if(this._storage) { return this._storage; }
    return this._storage = tmp
      .dir()
      .then(o => {
        this._tmpDirData = o;
        return o.path;
      });
  }

  /**
   * Spawn a new RethinkDB server instance, and wait for it to initialize.
   * @return {Promise<Number>} resolves with the server's listening port when the process has started.
   * @todo Retry with different ports if RethinkDB fails because of a taken port.
  */
  spawn() {
    if(this._spawn) { return this._spawn; }
    return this._spawn = Promise
      .all([
        this.port(),
        this.storage(),
      ])
      .then(([ { offset }, storage ]) => {
        this.spawned = spawn('rethinkdb', ['-o', `${offset}`, '-d', `${storage}`]);
        return this.bootWait(this.spawned).then(() => port);
      });
  }

  /**
   * Hack the `require("rethinkdb")` module to use the mocked server.
   * @return {Promise<RethinkDB>} resolves with the RethinkDB module after it's been modified
  */
  hackModule() {
    if(this._hackedModule) { return this._hackedModule; }
    return this._hackedModule = this
      .spawn()
      .then(port => {
        const r = require("rethinkdb");
        r.net.Connection.prototype.DEFAULT_PORT = port;
        return r;
      });
  }

  /**
   * Import all initial data into the mock database.
   * @return {Promise} resolves when all data is imported.
  */
  importData() {
    if(this._importedData) { return this._importedData; }
    if(!this.initialData) {
      return this._importedData = Promise.resolve();
    }
    return this._importedData = this
      .hackModule()
      .then(() => this.createDatabases())
      .then(() => this.createTables())
      .then(() => this.createDocuments());
  }

  /**
   * Cleans up from the mocked server.
  */
  cleanup() {
    if(this._cleanedup) { return this._cleanedup; }
    return this._cleanedup = this
      .shutdown()
      .then(() => this.removeStorage());
  }

  /**
   * Shuts down the RethinkDB instance.
   * @return {void}
   * @todo Check if actually running
  */
  shutdown() {
    this.spawned.kill();
  }

  /**
   * Cleanup the storage directory used for the database.
   * @return {Promise} resolves after the directory has been removed.
  */
  removeStorage() {
    if(this._removedStorage) { return this._removedStorage; }
    return this._removedStorage = Promise
      .all([
        this.storage(),
        this.shutdown(),
      ])
      .then(rimraf);
  }

  /**
   * Waits for the spawned RethinkDB process to fully initialize.
   * @param {child_process.spawn} spawned the spawned RethinkDB process
   * @return {Promise} resolves when the process has initialized.
   * @private
  */
  bootWait(spawned) {
    return new Promise((resolve, reject) => {
      spawned.stdout.on("data", chunk => {
        if(chunk.toString('utf8').startsWith('Listening for client driver connections')) {
          resolve();
        }
      });
    });
  }

  /**
   * Opens an active connection to the mocked database.
   * @return {Promise<Connection>} resolves with a RethinkDB connection.
  */
  connection() {
    if(this._connection) { return this._connection; }
    return this._connection = this
      .hackModule()
      .then(() => require("rethinkdb"));
  }

  /**
   * Creates the databases included in the initial data.
   * @return {Promise} resolves when all databases have been created.
   * @private
  */
  createDatabases() {
    if(this._createDatabases) { return this._createDatabases; }
    return this._createDatabases = Promise
      .all([
        this.hackModule(),
        this.connection(),
        Object.keys(this.initialDatabases()),
      ])
      .then(([ r, conn, dbs ]) => Promise.all( dbs.map(db => r.dbCreate(db).run(conn)) ));
  }

  /**
   * Create the tables included in the initial data.
   * @return {Promise} resolves when all tables have been created.
   * @private
  */
  createTables() {
    if(this._createTables) { return this._createTables; }
    return this._createTables = Promise
      .all([
        this.hackModule(),
        this.connection(),
        this.initialTables(),
        this.createDatabases(),
      ])
      .then(([ r, conn, tables ]) => Promise.all( tables.map(({ db, table }) => {
        return r.db(db).tableCreate(table).run(conn);
      })));
  }

  /**
   * Create the documents included in the initial data.
   * @return {Promise} resolves when all the documents have been created.
   * @private
  */
  createDocuments() {
    if(this._createDocs) { return this._createDocs; }
    return this._createDocs = Promise
      .all([
        this.hackModule(),
        this.connection(),
        this.initialTables(),
        this.createTables(),
      ])
      .then(([ r, conn, docs ]) => Promise.all( docs.map(({ db, table, doc }) => {
        return r.db(db).table(table).insert(doc).run(conn);
      })));
  }

  /**
   * Filters the initial databases to exclude test databases.
   * @return {Array<Object>} the valid databases.
   * @private
  */
  initialDatabases() {
    return filter(this.initialData, (val, index) => index !== "test");
  }

  /**
   * Reduces the tables from all of the inital databases into `[ { db: "name", table: "name", docs: [ ... ] } ]`.
   * @return {Array} each of the tables in the inital data
   * @private
  */
  initialTables() {
    function flattenDatabase(tables, db) {
      return map(tables, (docs, table) => { db, table, docs });
    }
    return flatMap(this.initialDatabases(), flattenDatabase);
  }

  /**
   * Reduces the documents from all tables in the initial databases into
   * ` [ { db: "name", table: "name", doc: { ... } } ]`
   * @return {Array} each of the tables in the initial data.
   * @private
  */
  initialDocuments() {
    function flattenTable({ db, table, docs }) {
      return map(docs, doc => { db, table, doc });
    }
    return flatMap(this.initialTables(), flattenTable);
  }

}

module.exports = MockedRethinkDB;


index.js
'use strict'

const { spawn } = require('child_process')
const rimraf = require('rimraf')
const mkdirp = require('mkdirp')

const getPortOffset = pid => {
  const maxOffset = 65535 - 28015
  return pid - (Math.floor(pid / maxOffset) * maxOffset)
}

const getStorageDir = (cwd, pid) => `${cwd}/.db-test-${pid}`

let rethink

module.exports.init = initialData => t => new Promise((resolve, reject) => {
  const offset = getPortOffset(process.pid)
  const port = 28015 + offset
  const r = require('rethinkdb')
  const dir = getStorageDir(process.cwd(), process.pid)

  r.net.Connection.prototype.DEFAULT_PORT = port
  mkdirp.sync(dir)
  rethink = spawn('rethinkdb', ['-o', `${offset}`, '-d', `${dir}/${port}`])

  if (process.env.AVA_RETHINKDB_DEBUG) {
    console.error(`==> Process ${process.pid} spawning RethinkDB server on port ${port}...`)
  }

  rethink.stdout.on('data', chunk => {
    if (chunk.toString('utf8').startsWith('Listening for client driver connections')) {
      if (initialData) {
        importData().then(() => resolve(port))
      } else {
        resolve(port)
      }

      if (process.env.AVA_RETHINKDB_DEBUG) {
        console.error(`==> Process ${process.pid} RethinkDB server booted!`)
      }
    }

    if (process.env.AVA_RETHINKDB_DEBUG) {
      console.error(chunk.toString())
    }
  })

  if (process.env.AVA_RETHINKDB_DEBUG) {
    rethink.stderr.on('data', chunk => console.error(chunk.toString()))
  }

  rethink.on('error', reject)

  function importData () {
    let conn
    return r.connect({}).then(_ => { conn = _ })
      .then(() => Promise.all(
        Object.keys(initialData).map(db => db !== 'test' && r.dbCreate(db).run(conn))
      ))
      .then(() => Promise.all(
        collectTables(initialData).map(([db, table]) => r.db(db).tableCreate(table).run(conn))
      ))
      .then(() => Promise.all(
        collectDocuments(initialData).map(([db, table, doc]) => r.db(db).table(table).insert(doc).run(conn))
      ))
  }
})

module.exports.cleanup = () => {
  if (process.env.AVA_RETHINKDB_DEBUG) {
    console.error(`==> Process ${process.pid} killing RethinkDB server...`)
  }

  rethink.kill()

  if (process.env.AVA_RETHINKDB_DEBUG) {
    console.error(`==> Process ${process.pid} killed RethinkDB server!`)
  }

  rimraf.sync(getStorageDir(process.cwd(), process.pid))
}

function collectTables (data) {
  return Object.keys(data)
               .map(db => Object.keys(data[db]).map(table => [db, table]))
               .reduce((a, b) => a.concat(b))
}

function collectDocuments (data) {
  return Object.keys(data)
               .map(db => Object.keys(data[db]).map(table => data[db][table].map(doc => [db, table, doc])))
               .reduce((a, b) => a.concat(b))
               .reduce((a, b) => a.concat(b))
}

module.exports.getPortOffset = getPortOffset
